### PR TITLE
Consents

### PR DESCRIPTION
**Pull request by antkiewiczk**

### CODE REVIEW

**Comment by tomaszdurka** [comment]
looked !

File: libs/consent/editor/src/lib/consent-list.tsx [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    })
+  }
+
+  return (
+    <>
+      <ListViewContainer>
+        <ListViewHeader>
+          <h2>{t('consents.title')}</h2>
+        </ListViewHeader>
+        {/* todo permission control */}
```
Comment (penk85) [inline libs/consent/editor/src/lib/consent-list.tsx:69]: Is this still to be done?

File: libs/consent/editor/src/lib/consentEditView.tsx [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  name: consent.name,
+  slug: consent.slug,
+  defaultValue: consent.defaultValue
+})
+
+const onErrorToast = (error: ApolloError, slug?: string) => {
+  if (error.message.includes('Unique constraint')) {
+    toaster.push(
+      <Message type="error" showIcon closable duration={3000}>
+        {`A consent with slug '${slug}' already exists. Please choose a different slug.`}
```
Comment (penk85) [inline libs/consent/editor/src/lib/consentEditView.tsx:29]: I'm guessing the error message isn't helpful here? Can we use translations for this?

File: libs/consent/editor/src/lib/consentForm.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            value={consent.slug ?? ''}
+            onChange={(slug: string) => onChange({slug})}
+          />
+        </Form.Group>
+
+        <Form.Group controlId="defaultValue">
+          <Form.ControlLabel>{t('consents.defaultValue')}</Form.ControlLabel>
+          <SelectPicker
+            key="default-value"
+            placeholder={t('userSubscriptionEdit.selectMemberPlan')}
```
Comment (penk85) [inline libs/consent/editor/src/lib/consentForm.tsx:60]: Did you want the member plan placeholder?

**Review by Itrulia — CHANGES_REQUESTED** [review]

**Review by tomaszdurka — CHANGES_REQUESTED** [review]
Remove `Unset` value from `ConsentValue`, change `ConsentValue` to simple `boolean`.

File: libs/consent/editor/rsuite.d.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import 'rsuite-table'
+
+declare module 'rsuite-table' {
+  type RowDataType<T = {[key: string]: any}> = {
+    dataKey?: string
+    children?: RowDataType[]
+  } & T
+}
```
Comment (tomaszdurka) [inline libs/consent/editor/rsuite.d.ts:8]: This cannot be loaded  from rsuite??

File: libs/consent/editor/src/apiClientv2.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import {ApolloClient, ApolloLink, createHttpLink, InMemoryCache} from '@apollo/client'
```
Comment (tomaszdurka) [inline libs/consent/editor/src/apiClientv2.ts:1]: Why we have this code in here? Do we need this in every module now?

Comment (tomaszdurka) [reply libs/consent/editor/src/apiClientv2.ts:1]: @Itrulia can you add your thoughts?

File: libs/api/prisma/schema.prisma [hunk 6]
```diff
@@ -809,6 +809,7 @@ model User {
   Invoice       Invoice[]
   CommentRating CommentRating[]
   PollVote      PollVote[]
+  UserConsent   UserConsent[]
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply libs/api/prisma/schema.prisma:812]: Yes we should, but that's out of scope for this PR

File: libs/api/prisma/schema.prisma [hunk 7]
```diff
@@ -984,3 +985,33 @@ model TaggedEvents {
...
+
+model Consent {
+  id         String   @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
+  createdAt  DateTime @default(now())
+  modifiedAt DateTime @updatedAt
+
+  name         String
+  slug         String        @unique
+  defaultValue Boolean
+  UserConsent  UserConsent[]
```
Comment (Itrulia) [inline libs/api/prisma/schema.prisma:997]:
```suggestion
  userConsents  UserConsent[]
```

File: libs/consent/api/src/lib/userConsent/user-consent.service.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+@Injectable()
+export class UserConsentService {
+  constructor(private prisma: PrismaClient) {}
+
+  /*
+  Queries
+ */
+  async userConsentList(filter?: UserConsentFilter): Promise<UserConsent[]> {
+    const data = await this.prisma.userConsent.findMany({
+      where: {
```
Comment (Itrulia) [inline libs/consent/api/src/lib/userConsent/user-consent.service.ts:20]:
```
where: {
   value: filter.value,
        consent: {
          name: filter.name,
          slug: filter.slug
        }
      },
```

Is probably what you wanted. Currently there's a bug here where value does nothing

File: libs/consent/api/src/lib/userConsent/user-consent.service.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    return data
+  }
+
+  /*
+  Mutations
+ */
+  async createUserConsent(userConsent: UserConsentInput, user: UserSession): Promise<UserConsent> {
+    // only allow creating for admin or affected user
+    if (!user.user.roleIDs.includes('admin') && user.user.id !== userConsent.userId) {
```
Comment (Itrulia) [inline libs/consent/api/src/lib/userConsent/user-consent.service.ts:59]: The service shouldn't be checking permissions but the resolver via decorators, this allows for things to be reused

**Review by Itrulia — APPROVED** [review]

**Review by tomaszdurka — APPROVED** [review]

**Pull request merged**